doctr/models/recognition/sar/pytorch.py

```diff
@@ -50,28 +50,35 @@ class AttentionModule(nn.Module):
 
     def __init__(self, feat_chans: int, state_chans: int, attention_units: int) -> None:
         super().__init__()
-        self.feat_conv = nn.Conv2d(feat_chans, attention_units, 3, padding=1)
+        self.feat_conv = nn.Conv2d(feat_chans, attention_units, kernel_size=3, padding=1)
         # No need to add another bias since both tensors are summed together
-        self.state_conv = nn.Linear(state_chans, attention_units, bias=False)
-        self.attention_projector = nn.Linear(attention_units, 1, bias=False)
+        self.state_conv = nn.Conv2d(state_chans, attention_units, kernel_size=1, bias=False)
+        self.attention_projector = nn.Conv2d(attention_units, 1, kernel_size=1, bias=False)
 
-    def forward(self, features: torch.Tensor, hidden_state: torch.Tensor) -> torch.Tensor:
-        # shape (N, C, H, W) -> (N, attention_units, H, W)
+    def forward(
+        self,
+        features: torch.Tensor,  # (N, C, H, W)
+        hidden_state: torch.Tensor,  # (N, C)
+    ) -> torch.Tensor:
+
+        height_feat_map, width_feat_map = features.size()[2:]
+
+        # (N, feat_chans, H, W) --> (N, attention_units, H, W)
         feat_projection = self.feat_conv(features)
-        # shape (N, L, rnn_units) -> (N, L, attention_units)
-        state_projection = self.state_conv(hidden_state).unsqueeze(-1).unsqueeze(-1)
-        # (N, L, attention_units, H, W)
-        projection = torch.tanh(feat_projection.unsqueeze(1) + state_projection)
-        # (N, L, H, W, 1)
-        attention = self.attention_projector(projection.permute(0, 1, 3, 4, 2))
-        # shape (N, L, H, W, 1) -> (N, L, H * W)
-        attention = torch.flatten(attention, 2)
-        attention = torch.softmax(attention, -1)
-        # shape (N, L, H * W) -> (N, L, 1, H, W)
-        attention = attention.reshape(-1, hidden_state.shape[1], features.shape[-2], features.shape[-1])
-
-        # (N, L, C)
-        return (features.unsqueeze(1) * attention.unsqueeze(2)).sum(dim=(3, 4))
+        # (N, state_chans, 1, 1) --> (N, attention_units, 1, 1)
+        hidden_state = hidden_state.view(hidden_state.size(0), hidden_state.size(1), 1, 1)
+        state_projection = self.state_conv(hidden_state)
+        state_projection = state_projection.expand(-1, -1, height_feat_map, width_feat_map)
+        # (N, attention_units, 1, 1) --> (N, attention_units, height_feat_map, width_feat_map)
+        attention_weights = torch.tanh(torch.add(feat_projection, state_projection, alpha=1))
+        # (N, attention_units, height_feat_map, width_feat_map) --> (N, 1, height_feat_map, width_feat_map)
+        attention_weights = self.attention_projector(attention_weights)
+        B, C, H, W = attention_weights.size()
+
+        # (N, H, W) --> (N, 1, H, W)
+        attention_weights = F.softmax(attention_weights.view(B, -1), dim=-1).view(B, C, H, W)
```

Comment:
I think it's recommended to use `torch.softmax` (they will deprecate F.softmax at some point if I recall correctly)